File: service/src/main/java/hu/bsstudio/robonaut/team/mapper/TeamModelEntityMapper.java
package hu.bsstudio.robonaut.team.mapper;

import hu.bsstudio.robonaut.entity.TeamEntity;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import java.awt.image.MultiPixelPackedSampleModel;
import lombok.AccessLevel;
import lombok.Setter;

public class TeamModelEntityMapper {

    @Setter(AccessLevel.PACKAGE)
    private ScoreEntityMapper scoreEntityMapper = new ScoreEntityMapper();

    public DetailedTeam toModel(final TeamEntity teamEntity) {
        return DetailedTeam.builder()
            .teamId(teamEntity.getTeamId())
            .year(teamEntity.getYear())
            .teamName(teamEntity.getTeamName())
            .teamMembers(teamEntity.getTeamMembers())
            .teamType(teamEntity.getTeamType())
            .skillScore(teamEntity.getSkillScore())
            .numberOfOvertakes(teamEntity.getNumberOfOvertakes())
            .safetyCarWasFollowed(teamEntity.isSafetyCarWasFollowed())
            .speedTimes(teamEntity.getSpeedTimes())
            .votes(teamEntity.getVotes())
            .audienceScore(teamEntity.getAudienceScore())
            .qualificationScore(teamEntity.getQualificationScore())
            .combinedScore(scoreEntityMapper.toModel(teamEntity.getScore()))
            .juniorScore(scoreEntityMapper.toModel(teamEntity.getJuniorScore()))
            .build();
    }

    public TeamEntity toEntity(final DetailedTeam detailedTeam) {
        final var entity = new TeamEntity();
        entity.setTeamId(detailedTeam.getTeamId());
        entity.setYear(detailedTeam.getYear());
        entity.setTeamName(detailedTeam.getTeamName());
        entity.setTeamMembers(detailedTeam.getTeamMembers());
        entity.setTeamType(detailedTeam.getTeamType());
        entity.setSkillScore(detailedTeam.getSkillScore());
        entity.setNumberOfOvertakes(detailedTeam.getNumberOfOvertakes());
        entity.setSafetyCarWasFollowed(detailedTeam.isSafetyCarWasFollowed());
        entity.setSpeedTimes(detailedTeam.getSpeedTimes());
        entity.setVotes(detailedTeam.getVotes());
        entity.setAudienceScore(detailedTeam.getAudienceScore());
        entity.setQualificationScore(detailedTeam.getQualificationScore());
        entity.setScore(scoreEntityMapper.toEntity(detailedTeam.getCombinedScore()));
        entity.setJuniorScore(scoreEntityMapper.toEntity(detailedTeam.getJuniorScore()));
        return entity;
    }
}

File: service/src/main/kotlin/hu/bsstudio/robonaut/team/mapper/TeamModelEntityMapper.kt
package hu.bsstudio.robonaut.team.mapper

import hu.bsstudio.robonaut.entity.TeamEntity
import hu.bsstudio.robonaut.team.model.DetailedTeam

class TeamModelEntityMapper(private val scoreEntityMapper: ScoreEntityMapper = ScoreEntityMapper()) {

    fun toModel(teamEntity: TeamEntity) = DetailedTeam(
        teamId = teamEntity.teamId,
        year = teamEntity.year,
        teamName = teamEntity.teamName,
        teamMembers = teamEntity.teamMembers,
        teamType = teamEntity.teamType,
        skillScore = teamEntity.skillScore,
        numberOfOvertakes = teamEntity.numberOfOvertakes,
        safetyCarWasFollowed = teamEntity.safetyCarWasFollowed,
        speedTimes = teamEntity.speedTimes,
        votes = teamEntity.votes,
        audienceScore = teamEntity.audienceScore,
        qualificationScore = teamEntity.qualificationScore,
        combinedScore = scoreEntityMapper.toModel(teamEntity.score),
        juniorScore = scoreEntityMapper.toModel(teamEntity.juniorScore),
    )

    fun toEntity(detailedTeam: DetailedTeam) = TeamEntity(
        teamId = detailedTeam.teamId,
        year = detailedTeam.year,
        teamName = detailedTeam.teamName,
        teamMembers = detailedTeam.teamMembers,
        teamType = detailedTeam.teamType,
        skillScore = detailedTeam.skillScore,
        numberOfOvertakes = detailedTeam.numberOfOvertakes,
        safetyCarWasFollowed = detailedTeam.safetyCarWasFollowed,
        speedTimes = detailedTeam.speedTimes,
        votes = detailedTeam.votes,
        audienceScore = detailedTeam.audienceScore,
        qualificationScore = detailedTeam.qualificationScore,
        score = scoreEntityMapper.toEntity(detailedTeam.combinedScore),
        juniorScore = scoreEntityMapper.toEntity(detailedTeam.juniorScore),
    )
}
